### PR TITLE
[FEATURE] Implement Feature Importance in Matching Tasks for HypEx

### PR DESCRIPTION
**Pull request by tikhomirovd**
### Description

Implemented feature selection in HypEx, specifically addressing the need for feature importance in matching tasks. This enhancement not only restores a capability that was present in earlier versions (as an add-on for LAMA) but also significantly improves HypEx's standalone analytical power.

### Changes Made

- Integrated feature importance algorithms using CatBoost and LightGBM.
- Developed a mechanism to tie feature importance directly into the matching tasks within HypEx.
- Added documentation and examples to guide users in leveraging this new feature.
- Conducted extensive code reviews and refactoring to ensure seamless integration.

### Related Issues

- 🚀 Feature Proposal: [Implement Feature Importance for Matching Task in HypEx #12 ](https://github.com/sb-ai-lab/HypEx/issues/12)

### Additional Notes

- The implementation focuses on compatibility with current matching algorithms in HypEx.
- Reviewers are encouraged to pay special attention to the integration points and the potential impacts on existing functionalities.

### Testing and Validation

- Conducted rigorous testing with various datasets to validate the accuracy and reliability of the feature importance results.
- Ensured compatibility and performance with existing matching tasks.

### Performance Considerations

- Evaluated the performance impacts, particularly in terms of computation time and resource utilization.
- Optimization efforts were made to ensure minimal performance degradation.

### Breaking Changes

- No significant breaking changes were introduced. Backward compatibility with existing functionalities was maintained.

### Dependencies

- Added dependencies include `catboost` and `lightgbm` for the underlying feature importance calculations.

### Merge Request Checklist

- [x] Code follows project coding guidelines.
- [x] Documentation reflects the changes made.
- [x] Unit tests cover new or changed functionality.
- [x] Performance and breaking changes have been considered.

### CODE REVIEW

File: hypex/selectors/feature_selector.py [hunk 1]
```diff
@@ -21,50 +22,22 @@ class FeatureSelector:
...
-            n_folds:
-                Number of folds for cross-validation
-            verbose:
-                Flag to control the verbosity of the process stages
-            generate_report:
-                Flag to control whether to create a report or not
-            report_dir:
-                Directory for storing report files
             use_algos:
                 List of names of LAMA algorithms for feature selection
```
Comment (VasilievaKA) [inline hypex/selectors/feature_selector.py:36]: Are these still LAMA algorithms?

File: hypex/selectors/selector_primal_methods.py [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+        info_col_list:
+        target:
+        treatment:
+        weights_col_list:
+        category_col_list:
+
+    Returns: marked column sets and copy of processed dataframe
+
+    """
+    if info_col_list is None:
```
Comment (VasilievaKA) [inline hypex/selectors/selector_primal_methods.py:50]: info_col_list = [] if info_col_list is None else info_col_list 
maybe like this?

File: hypex/selectors/selector_primal_methods.py [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+
+        feature_importance = get_feature_importance_df(
+            lgbm_selection_model, feature_col_list, _target_col)
+        feature_importance_list.append(feature_importance)
+
+    report_df = concat_feature_importance_dfs(feature_importance_list)
+    return report_df
+
+
+def pd_catboost_feature_selector(
```
Comment (VasilievaKA) [inline hypex/selectors/selector_primal_methods.py:170]: Maybe add **kwargs for all methods? 
Users will be able to configure the regressors in their own way

File: hypex/matcher.py [hunk 4]
```diff
@@ -165,6 +165,8 @@ def __init__(
             pbar:
                 Display progress bar while get index
         """
+        self.short_features_df = None
+        self.detailed_features_df = None
         if use_algos is None:
```
Comment (VasilievaKA) [inline hypex/matcher.py:170]: use_algos = USE_ALGOS if use_algos is None else use_algos

File: hypex/matcher.py [hunk 5]
```diff
@@ -335,42 +337,31 @@ def match_no_rep(self, threshold: float = 0.1, approximate_match: bool = False)
                                       self.input_data.loc[np.concatenate(filtred_matches.values)]])
         return matched_data
 
-    def lama_feature_select(self) -> pd.DataFrame:
+    def feature_select(self) -> pd.DataFrame:
         """Calculates the importance of each feature.
 
         This method use LamaFeatureSelector to rank the importance of each feature in the dataset
```
Comment (VasilievaKA) [inline hypex/matcher.py:343]: LamaFeatureSelector?

File: hypex/selectors/feature_selector.py [hunk 6]
```diff
@@ -3,6 +3,7 @@
 from typing import List
 
 import pandas as pd
+from hypex.selectors import selector_primal_methods
 
 logger = logging.getLogger("lama_feature_selector")
```
Comment (VasilievaKA) [inline hypex/selectors/feature_selector.py:8]: lama_feature_selector?

File: hypex/selectors/feature_selector.py [hunk 7]
```diff
@@ -21,50 +22,22 @@ class FeatureSelector:
...
-            timeout: int,
-            n_threads: int,
-            n_folds: int,
-            verbose: bool,  # не используется
-            generate_report: bool,
-            report_dir: str,
-            use_algos: List[str],
+            use_algos: str,
     ):
         """Initialize the LamaFeatureSelector.
```
Comment (VasilievaKA) [inline hypex/selectors/feature_selector.py:28]: LamaFeatureSelector?

File: hypex/selectors/feature_selector.py [hunk 8]
```diff
@@ -84,19 +57,21 @@ def perform_selection(self, df: pd.DataFrame) -> pd.DataFrame:
...
+            selector = selector_primal_methods.pd_ridgecv_feature_selector
         else:
-            task_name = "multiclass"
-            loss = "crossentropy"
-            metric = "crossentropy"
-
-        features_scores = []
-
-        return features_scores
+            raise Exception(f"Unknown input algorithm used on feature_selector: {self.use_algos}")
```
Comment (VasilievaKA) [inline hypex/selectors/feature_selector.py:67]: Maybe it would be better to assign USE_ALGOS when the name of the algorithm is entered incorrectly? And throw out a warning
Otherwise, the whole pipeline will collapse in the future.

File: hypex/selectors/selector_primal_methods.py [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+        treatment:
+        weights_col_list:
+        category_col_list:
+
+    Returns: marked column sets and copy of processed dataframe
+
+    """
+    if info_col_list is None:
+        info_col_list = []
+    if weights_col_list is None:
```
Comment (VasilievaKA) [inline hypex/selectors/selector_primal_methods.py:52]: And below too

File: hypex/matcher.py (hunk 5 above)
Comment (tikhomirovd) [reply hypex/matcher.py:343]: it was lama, for now not lama

File: hypex/matcher.py [hunk 10]
```diff
@@ -549,7 +540,11 @@ def estimate(self, features: list = None) -> tuple:
             Results of matching and matching quality metrics
         """
         if features is not None:
+            features = [*features]
```
Comment (Dmatryus) [inline hypex/matcher.py:543]: Replace `list` with `Iterable` in arguments.

File: hypex/selectors/feature_selector.py (hunk 8 above)
Comment (Dmatryus) [reply hypex/selectors/feature_selector.py:67]: In `__init__` for Example.

Comment (Dmatryus) [reply hypex/selectors/feature_selector.py:67]: And get selectors from dict.

File: hypex/selectors/selector_primal_methods.py [hunk 11]
```diff
@@ -0,0 +1,334 @@
...
+    """
+    if col_list:
+        dtype = 'str' if is_category else 'float'
+        stats = 'top' if is_category else '50%'
+
+        fill_values = df[col_list].astype(dtype).describe(include='all').loc[stats]
+        df[col_list] = df[col_list].fillna(fill_values)
+
+
+def pd_input_preproc(
```
Comment (Dmatryus) [inline hypex/selectors/selector_primal_methods.py:45]: Maybe `input_preprocessing` is better?

File: hypex/matcher.py (hunk 10 above)
Comment (Dmatryus) [inline hypex/matcher.py:543]: Replace `list` with `Iterable` in arguments.

File: hypex/selectors/feature_selector.py (hunk 8 above)
Comment (Dmatryus) [reply hypex/selectors/feature_selector.py:67]: Check this in `__init__` and from `dict` Instead of `if-else`.

File: hypex/selectors/selector_primal_methods.py (hunk 11 above)
Comment (Dmatryus) [inline hypex/selectors/selector_primal_methods.py:45]: Maybe `input_preprocessing` is better?

File: hypex/selectors/selector_primal_methods.py [hunk 12]
```diff
@@ -0,0 +1,334 @@
...
+    df = df.copy(deep=True)
+
+    feature_col_list = [
+        col for col in df.columns
+        if col not in info_col_list + target_col_list + treatment_col_list + weights_col_list
+    ]
+
+    numeric_col_list = [col for col in feature_col_list if col not in category_col_list]
+
+    df[category_col_list] = df[category_col_list].astype('str')
```
Comment (Dmatryus) [inline hypex/selectors/selector_primal_methods.py:87]: Both fillna in one func

File: hypex/selectors/selector_primal_methods.py (hunk 3 above)
Comment (tikhomirovd) [reply hypex/selectors/selector_primal_methods.py:221]: Will implement in the next version

**Review by Dmatryus — APPROVED** [review]
GG!

**Comment by tikhomirovd** [comment]
Done, great job!
